source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,412 @@
+  }
+}
+
+Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                          bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isAcceptEncodingAllowed(headers) && isContentTypeAllowed(headers) &&
+      !hasCacheControlNoTransform(headers) && isEtagAllowed(headers) &&
+      isTransferEncodingAllowed(headers) && !headers.ContentEncoding()) {
+    sanitizeEtagHeader(headers);
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.setContentEncoding(config_->contentEncoding());
+    config_->stats().compressed_.inc();
+    compressor_ = config_->makeCompressor();
+  } else if (!skip_compression_) {
+    skip_compression_ = true;
+    config_->stats().not_compressed_.inc();
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (!skip_compression_) {
+    config_->stats().total_uncompressed_bytes_.add(data.length());
+    compressor_->compress(data, end_stream ? Compressor::State::Finish : Compressor::State::Flush);
+    config_->stats().total_compressed_bytes_.add(data.length());
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus CompressorFilter::encodeTrailers(Http::ResponseTrailerMap&) {
+  if (!skip_compression_) {
+    Buffer::OwnedImpl empty_buffer;
+    compressor_->compress(empty_buffer, Compressor::State::Finish);
+    config_->stats().total_compressed_bytes_.add(empty_buffer.length());
+    encoder_callbacks_->addEncodedData(empty_buffer, true);
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+bool CompressorFilter::hasCacheControlNoTransform(Http::ResponseHeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                     Http::Headers::get().CacheControlValues.NoTransform);
+  }
+
+  return false;
+}
+
+std::unique_ptr<CompressorFilter::EncodingDecision>
+CompressorFilter::chooseEncoding(const Http::ResponseHeaderMap& headers) const {
+  using EncPair = std::pair<absl::string_view, float>; // pair of {encoding, q_value}
+  std::vector<EncPair> pairs;
+  absl::string_view content_type_value;
+
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr) {
+    content_type_value =
+        StringUtil::trim(StringUtil::cropRight(content_type->value().getStringView(), ";"));
+  }
+
+  // There could be many compressors registered for the same content encoding, e.g. consider a case
+  // when there are two gzip filters using different compression levels for different content sizes.
+  // In such case we ignore duplicates (or different filters for the same encoding) registered last.
```

Comment:
In particular this method needs massively more comments, both at the high level and throughout. Please describe in detail how filter state is used, how things interact, etc.